### PR TITLE
Tests fix

### PR DESCRIPTION
**Pull request by tkornuta-nvidia**
- [x] fixes serialization unit tests
- [x] refactors all tests to rely on PyTest (!)
- [x] fixes test_save_restore_from_nemo_file test
- [x] actually fixes the model restoring 

IMO this is not the final solution - but at least we can start exporting models to JARVIS...

### CODE REVIEW

File: tests/collections/asr/test_asr_ctcencdec_model.py [hunk 1]
```diff
@@ -89,10 +87,12 @@ def test_constructor(self):
             {'preprocessor': DictConfig(preprocessor), 'encoder': DictConfig(encoder), 'decoder': DictConfig(decoder)}
         )
         asr_model = EncDecCTCModel(cfg=modelConfig)
+        # WHAT IS THAT?!
         asr_model.train()
```
Comment (titu1994) [inline tests/collections/asr/test_asr_ctcencdec_model.py:91]: ... we can "train" without a scheduler or Optimizer? @okuchaiev

File: tests/core/test_asr_metrics.py [hunk 2]
```diff
@@ -119,8 +117,11 @@ def get_wer(prediction: str, reference: str):
             n2 = random.randint(0, 512)
             s1 = __randomString(n1)
             s2 = __randomString(n2)
-            logging.debug(n1)
-            logging.debug(s1)
-            logging.debug(n2)
-            logging.debug(s2)
-            self.assertEqual(get_wer(prediction=s1, reference=s2), word_error_rate(hypotheses=[s1], references=[s2]))
+
+            assert get_wer(prediction=s1, reference=s2) == word_error_rate(hypotheses=[s1], references=[s2])
```
Comment (titu1994) [inline tests/core/test_asr_metrics.py:122]: Please do a range check here instead of a hardcoded == so we can account for this unstable behaviour. Even better if we can remove this instability entirely.

Comment (titu1994) [reply tests/core/test_asr_metrics.py:122]: @tkornuta-nvidia For now please the mark the test with pleasefixme mark so that @okuchaiev can take a look.

File: tests/collections/asr/test_asr_ctcencdec_model.py (hunk 1 above)
Comment (titu1994) [reply tests/collections/asr/test_asr_ctcencdec_model.py:91]: @tkornuta-nvidia For now, maybe revert this line and remove your comment

File: tests/core/test_asr_metrics.py [hunk 3]
```diff
@@ -119,8 +118,19 @@ def get_wer(prediction: str, reference: str):
             n2 = random.randint(0, 512)
             s1 = __randomString(n1)
             s2 = __randomString(n2)
-            logging.debug(n1)
-            logging.debug(s1)
-            logging.debug(n2)
-            logging.debug(s2)
```
Comment (titu1994) [inline tests/core/test_asr_metrics.py:125]: Please leave these debug statements for @okuchaiev to review, they were added by @blisc

File: tests/core/test_fileio.py [hunk 4]
```diff
@@ -21,80 +21,85 @@
...
+class FileIOTest:
+    @pytest.mark.unit
+    def test_to_from_config_file(self, asr_model):
 
         with tempfile.NamedTemporaryFile() as fp:
             yaml_filename = fp.name
             asr_model.to_config_file(path2yaml_file=yaml_filename)
             next_instance = EncDecCTCModel.from_config_file(path2yaml_file=yaml_filename)
-            self.assertTrue(isinstance(next_instance, EncDecCTCModel))
+            assert isinstance(next_instance, EncDecCTCModel) == True
```
Comment (titu1994) [inline tests/core/test_fileio.py:102]: Why is the == True required?

File: tests/collections/asr/test_asr_ctcencdec_model.py (hunk 1 above)
Comment (tkornuta-nvidia) [reply tests/collections/asr/test_asr_ctcencdec_model.py:91]: I honestly do not know what is happening here... moreover, IF there is training, then it is clearly not unit tests...

File: tests/core/test_asr_metrics.py (hunk 3 above)
Comment (tkornuta-nvidia) [reply tests/core/test_asr_metrics.py:125]: Logging shouldn't be part of any test.

File: tests/core/test_asr_metrics.py (hunk 2 above)
Comment (tkornuta-nvidia) [reply tests/core/test_asr_metrics.py:122]: Actually I haven't _touched_ the `@pleasefixme` decorator... Seems that @blisc was playing with it too, then I had to solve conflicts and it was gone. Afterwards I have reverted it.

File: tests/core/test_fileio.py (hunk 4 above)
Comment (tkornuta-nvidia) [reply tests/core/test_fileio.py:102]: correct, it is redundant

Comment (tkornuta-nvidia) [reply tests/core/test_fileio.py:102]: In the other places I have fixed it correctly, speed of light, you know ;)

**Comment by tkornuta-nvidia** [comment]
From what I am seeing the current restore_from simply cannot work as it is using model.load_from_checkpoint()
https://github.com/NVIDIA/NeMo/blob/290a6018217b98f0d6bc4f308f84668f04e4147d/nemo/core/classes/modelPT.py#L126
by calling:
```
instance = cls.load_from_checkpoint(model_weights, cfg=conf)
```
whereas the signature of the methods is:
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/saving.py#L47
```
    @classmethod
    def load_from_checkpoint(
            cls,
            checkpoint_path: str,
            *args,
            map_location: Optional[Union[Dict[str, str], str, torch.device, int, Callable]] = None,
            hparams_file: Optional[str] = None,
            tags_csv: Optional[str] = None,  # backward compatible, todo: remove in v0.9.0
            **kwargs
    ):
```

Sadly, I cannot finish this right now, will work on this on Monday.

**Comment by tkornuta-nvidia** [comment]
Btw. The CI is working - the test is still denoted at `@pleasefixme`

**Comment by tkornuta-nvidia** [comment]
My final comment on this is as follows:

@okuchaiev seems that in `model.save_to()` you want not only to save the *model weights, but also training state*, is it correct?

If so, then the PTL trainer has a different method to load that:
https://pytorch-lightning.readthedocs.io/en/latest/weights_loading.html
```
trainer = Trainer(resume_from_checkpoint='some/path/to/my_checkpoint.ckpt')
```

but please note - this method instantiates *the trainer, not the model*.

File: tests/core/test_asr_metrics.py (hunk 3 above)
Comment (blisc) [reply tests/core/test_asr_metrics.py:125]: This test is randomly and sometimes randomly fails. That’s why the logging is there so we can catch the corner cases where it fails

File: tests/collections/asr/test_asr_ctcencdec_model.py (hunk 1 above)
Comment (okuchaiev) [reply tests/collections/asr/test_asr_ctcencdec_model.py:91]: This is just a call of  torch.nn.Module.train() method.

File: tests/core/test_asr_metrics.py (hunk 2 above)
Comment (okuchaiev) [reply tests/core/test_asr_metrics.py:122]: Need to think about this one carefully - there is no randomness here. The mismatch could be due to floating operations order only, so it either matches exactly or the range should be pretty tight. Please leave as == for now

File: tests/core/test_asr_metrics.py [hunk 5]
```diff
@@ -119,8 +118,19 @@ def get_wer(prediction: str, reference: str):
...
-            self.assertEqual(get_wer(prediction=s1, reference=s2), word_error_rate(hypotheses=[s1], references=[s2]))
+
+            assert get_wer(prediction=s1, reference=s2) == word_error_rate(hypotheses=[s1], references=[s2])
+
+            # @Oleksii: THIS TEST IS UNSTABLE!
+
+            # One failed case:
```
Comment (okuchaiev) [inline tests/core/test_asr_metrics.py:126]: this is useless information - please remove

Comment (okuchaiev) [reply tests/core/test_asr_metrics.py:126]: need actual s1 and s2 to see what's going on

File: tests/core/test_asr_metrics.py (hunk 3 above)
Comment (blisc) [reply tests/core/test_asr_metrics.py:125]: Please keep these logging messages. Yes I understand logging should not be a part of any **working test**. The caveat here is that this test is not working as you have mentioned in your comments. So please keep these so it can log s1 and s2 in the cases where the tests fail. That way we can debug the root causes of these failures.

File: tests/collections/asr/test_asr_ctcencdec_model.py (hunk 1 above)
Comment (tkornuta-nvidia) [reply tests/collections/asr/test_asr_ctcencdec_model.py:91]: eheheh, train()/eval()! right... ;)

File: tests/core/test_asr_metrics.py (hunk 3 above)
Comment (okuchaiev) [reply tests/core/test_asr_metrics.py:125]: yes, test fails occasionally - I need that info when it fails to diagnose.

**Comment by okuchaiev** [comment]
thanks @tkornuta-nvidia  for cleaning up unittest/pytest mess

File: tests/collections/asr/test_asr_ctcencdec_model.py [hunk 6]
```diff
@@ -89,10 +87,12 @@ def test_constructor(self):
             {'preprocessor': DictConfig(preprocessor), 'encoder': DictConfig(encoder), 'decoder': DictConfig(decoder)}
         )
         asr_model = EncDecCTCModel(cfg=modelConfig)
+        # WHAT IS THAT?!
```
Comment (blisc) [inline tests/collections/asr/test_asr_ctcencdec_model.py:90]: Remove comment

File: tests/core/test_asr_metrics.py [hunk 7]
```diff
@@ -86,14 +86,15 @@ def get_wer(prediction: str, reference: str):
...
-        self.assertEqual(get_wer('a f c', 'a b c'), 1.0 / 3.0)
+        assert get_wer('cat', 'cot') == 1.0
+        assert get_wer('gpu', 'g p u') == 1.0
+        assert get_wer('g p u', 'gpu') == 3.0
+        assert get_wer('ducati motorcycle', 'motorcycle') == 1.0
+        assert get_wer('ducati motorcycle', 'ducuti motorcycle') == 0.5
+        assert get_wer('a f c', 'a b c') == 1.0 / 3.0
 
     @pytest.mark.unit
+    @pytest.mark.pleasefixme
```
Comment (blisc) [inline tests/core/test_asr_metrics.py:97]: Can we unmark for now? Let it error out and address it then?

File: tests/collections/asr/test_asr_ctcencdec_model.py (hunk 6 above)
Comment (tkornuta-nvidia) [reply tests/collections/asr/test_asr_ctcencdec_model.py]: removed

File: nemo/core/classes/common.py [hunk 8]
```diff
@@ -81,16 +82,21 @@ def _attach_and_validate_output_types(self, out_objects):
...
+class Serialization:
     @classmethod
     def from_config_dict(cls, config: DictConfig):
         """Instantiates object using DictConfig-based configuration"""
+
         if 'cls' in config and 'params' in config:
-            # regular hydra-based instantiation
-            instance = hydra.utils.instantiate(config=config)
+            class_obj = eval(config.cls)
+            instance = class_obj(cfg=config)
```
Comment (tkornuta-nvidia) [inline nemo/core/classes/common.py:92]: Please note that this solution currently work only for Classes that have "cls" in their __init__.

I have few ideas here - the question is how we generally envision serialization/deserialization. We need to sync on this.

File: tests/core/test_fileio.py [hunk 9]
```diff
@@ -21,161 +21,116 @@
...
+
+            assert len(asr_model.decoder.vocabulary) == len(asr_model2.decoder.vocabulary)
+            assert asr_model.num_weights == asr_model2.num_weights
+
             w1 = asr_model.encoder.encoder[0].mconv[0].conv.weight.data.detach().cpu().numpy()
             w2 = asr_model2.encoder.encoder[0].mconv[0].conv.weight.data.detach().cpu().numpy()
-            self.assertTrue(np.array_equal(w1, w2))
+
+            print(w1)
+            print(w2)
```
Comment (tkornuta-nvidia) [inline tests/core/test_fileio.py:135]: Ech, missed those two. Still, pytest won't print them (unless an error will pop out).

Will fix this in next PR - where I will try to unify the serialization/checkpointing for modules/models with pure cfg and list of arguments.

File: tests/core/test_asr_metrics.py (hunk 7 above)
Comment (tkornuta-nvidia) [reply tests/core/test_asr_metrics.py:96]: right! thanks, forgot about this...

Comment (tkornuta-nvidia) [reply tests/core/test_asr_metrics.py:96]: done

File: tests/core/test_fileio.py (hunk 9 above)
Comment (tkornuta-nvidia) [reply tests/core/test_fileio.py]: fixed

**Comment by tkornuta-nvidia** [comment]
```
(nemo-env) tkornuta@tkornuta-desktop:~/workspace/nemo$ pytest -m unit
=========================================================================================== test session starts ============================================================================================
platform linux -- Python 3.7.4, pytest-5.3.5, py-1.8.1, pluggy-0.13.1 -- /home/tkornuta/anaconda3/envs/nemo-env/bin/python
cachedir: .pytest_cache
rootdir: /home/tkornuta/workspace/nemo, inifile: setup.cfg
collected 20 items

tests/collections/asr/test_asr_ctcencdec_model.py::TestEncDecCTCModel::test_constructor PASSED                                                                                                       [  5%]
tests/collections/asr/test_asr_modules.py::TestASRModulesBasicTests::test_AudioToMelSpectrogramPreprocessor PASSED                                                                                   [ 10%]
tests/collections/asr/test_asr_modules.py::TestASRModulesBasicTests::test_SpectrogramAugmentationr PASSED                                                                                            [ 15%]
tests/core/test_asr_metrics.py::TestWordErrorRate::test_wer_function PASSED                                                                                                                          [ 20%]
tests/core/test_asr_metrics.py::TestWordErrorRate::test_wer_metric_simple PASSED                                                                                                                     [ 25%]
tests/core/test_asr_metrics.py::TestWordErrorRate::test_wer_metric_randomized PASSED                                                                                                                 [ 30%]
tests/core/test_fileio.py::TestFileIO::test_to_from_config_file PASSED                                                                                                                               [ 35%]
tests/core/test_fileio.py::TestFileIO::test_save_restore_from_nemo_file PASSED                                                                                                                       [ 40%]
tests/core/test_neural_types.py::TestNeuralTypeSystem::test_short_vs_long_version PASSED                                                                                                             [ 45%]
tests/core/test_neural_types.py::TestNeuralTypeSystem::test_parameterized_type_audio_sampling_frequency PASSED                                                                                       [ 50%]
tests/core/test_neural_types.py::TestNeuralTypeSystem::test_transpose_same_1 PASSED                                                                                                                  [ 55%]
tests/core/test_neural_types.py::TestNeuralTypeSystem::test_transpose_same_2 PASSED                                                                                                                  [ 60%]
tests/core/test_neural_types.py::TestNeuralTypeSystem::test_inheritance_spec_augment_example PASSED                                                                                                  [ 65%]
tests/core/test_neural_types.py::TestNeuralTypeSystem::test_singletone PASSED                                                                                                                        [ 70%]
tests/core/test_neural_types.py::TestNeuralTypeSystem::test_list_of_lists PASSED                                                                                                                     [ 75%]
tests/core/test_neural_types.py::TestNeuralTypeSystem::test_void PASSED                                                                                                                              [ 80%]
tests/core/test_neural_types.py::TestNeuralTypeSystem::test_big_void PASSED                                                                                                                          [ 85%]
tests/core/test_neural_types.py::TestNeuralTypeSystem::test_unspecified_dimensions PASSED                                                                                                            [ 90%]
tests/core/test_neural_types.py::TestNeuralTypeSystem::test_any_axis PASSED                                                                                                                          [ 95%]
tests/core/test_neural_types.py::TestNeuralTypeSystem::test_struct PASSED                                                                                                                            [100%]
```

File: nemo/core/classes/common.py (hunk 8 above)
Comment (okuchaiev) [reply nemo/core/classes/common.py:92]: OK, will do outside this PR

File: nemo/collections/asr/models/ctc_models.py [hunk 10]
```diff
@@ -37,17 +44,23 @@ class EncDecCTCModel(ASRModel):
...
             # This is for Jarvis service. Adding here for now to avoid effects of decorators
-            OmegaConf.set_struct(cfg, False)
-            cfg.cls = 'nemo.collections.asr.models.EncDecCTCModel'
-            OmegaConf.set_struct(cfg, True)
+            # Assuming cfg will have the config structure CONSISTENT with the modules, i.e.: {cls: ..., params: ...}
+            config = {"cls": 'nemo.collections.asr.models.EncDecCTCModel'}
+            if "params" in cfg:
+                config["params"] = cfg.params
+            else:
+                config["params"] = cfg
```
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:52]: Why are we pushing the entire cfg into the "params" namespace here? This will break optimizers and schedulers cause they assume everything is run under the model.optim namespace.

File: nemo/core/classes/common.py [hunk 11]
```diff
@@ -81,16 +82,21 @@ def _attach_and_validate_output_types(self, out_objects):
...
-class Serialization(ABC):
+class Serialization:
     @classmethod
     def from_config_dict(cls, config: DictConfig):
         """Instantiates object using DictConfig-based configuration"""
+
         if 'cls' in config and 'params' in config:
-            # regular hydra-based instantiation
-            instance = hydra.utils.instantiate(config=config)
+            class_obj = eval(config.cls)
```
Comment (titu1994) [inline nemo/core/classes/common.py:91]: Raw eval to inflate class ?? This is generally considered highly dangerous behavior throughout python. You can inject any class string here.

File: nemo/core/classes/modelPT.py [hunk 12]
```diff
@@ -90,13 +91,14 @@ def save_to(self, save_path: str):
...
-            if trainer.model != self:
-                trainer.model = self
-            trainer.save_checkpoint(filepath=model_weights, weights_only=True)
+            # Prepare filenames.
+            config_yaml_file = path.join(tmpdir, _MODEL_CONFIG_YAML)
+            model_weights_file = path.join(tmpdir, _MODEL_WEIGHTS)
+            # Save the configuration.
+            self.to_config_file(path2yaml_file=config_yaml_file)
+            # Save the weights.
+            save(self.state_dict(), model_weights_file)
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:100]: Please use the explicit torch.save() here to have semantic understanding of which "save" we are actually calling.

File: nemo/core/classes/modelPT.py [hunk 13]
```diff
@@ -120,10 +122,21 @@ def restore_from(cls, restore_path: str):
...
+
+            # Load the configuration.
+            config = OmegaConf.load(config_yaml_file)
+            print(config)
+
+            # Instantiate the object.
+            instance = Serialization.from_config_dict(config)
+
+            # Load the weights.
+            state_dict = load(model_weights_file, map_location=lambda storage, loc: storage)
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:137]: Please use explicit `torch.load()` here to have semantic understanding of which load we are calling here.

File: nemo/collections/asr/models/ctc_models.py (hunk 10 above)
Comment (titu1994) [reply nemo/collections/asr/models/ctc_models.py:52]: On top of this, there is no good reason to push everything into config. You could easily simply add the cfg.cls as I did above with struct overriding and treat the rest of the namespace as part of `cfg` itself - this is easier to interpret, and still serves the same purpose.

File: nemo/core/classes/modelPT.py (hunk 13 above)
Comment (okuchaiev) [reply nemo/core/classes/modelPT.py:137]: this is wrong - should not be using torch.load(). This PR is fixing tests - I'll work on save/restore logic

Comment (tkornuta-nvidia) [reply nemo/core/classes/modelPT.py:137]: I didn't wanted to user torch.load, as I didn't wanted to call `import torch`

File: nemo/core/classes/common.py (hunk 11 above)
Comment (tkornuta-nvidia) [reply nemo/core/classes/common.py:91]: @titu1994 you are right, I know, this is far from perfect, just wanted to a) have "save/restore" working and b) unblock our CI ASAP - and I spent more than a day working on this already.

I can work on this as part of Serialization PR - I will have to change plenty of things in here to make it work with graphs anyway...

File: nemo/collections/asr/models/ctc_models.py (hunk 10 above)
Comment (tkornuta-nvidia) [reply nemo/collections/asr/models/ctc_models.py:52]: Honestly, I did this change just to be consistent, as described in the comment above:
https://github.com/NVIDIA/NeMo/blob/c7988ef531fc0092004e516347feac17ff4e62a4/nemo/collections/asr/models/ctc_models.py#L47

I have went through @okuchaiev examples/tests and he assumed that his:
https://github.com/NVIDIA/NeMo/blob/c7988ef531fc0092004e516347feac17ff4e62a4/tests/collections/asr/test_asr_ctcencdec_model.py#L25

So I just followed. It doesn't even mean that I agree with this - I have just followed the same pattern...

File: nemo/core/classes/common.py (hunk 11 above)
Comment (titu1994) [reply nemo/core/classes/common.py:91]: Tom, there was nothing blocking CI - all tests were running, albeit using unittests rather than pytests. Oleksii said he would fix serialization, but you have gone ahead and merged this solution. At this point I dont understand how this PR got approved with this kind of class instantiation, and I just hope future PR fixes this as a high priority patch.

File: nemo/collections/asr/models/ctc_models.py (hunk 10 above)
Comment (titu1994) [reply nemo/collections/asr/models/ctc_models.py:52]: Yes and that pattern doesnt work with all of optim namespace. I dont understand why a PR to refactor tests from unittest to pytest has gone ahead and broken a good chunk of other stuff unrelated to tests.

Comment (tkornuta-nvidia) [reply nemo/collections/asr/models/ctc_models.py:52]: @titu1994 regarding optimizers: this is important, are you sure?

I have just relied on our tests, and _believed_ that if tests are passing, then I am not breaking anything:
http://10.110.42.104:8080/blue/organizations/jenkins/NeMo/detail/PR-871/12/pipeline

File: nemo/core/classes/common.py (hunk 11 above)
Comment (tkornuta-nvidia) [reply nemo/core/classes/common.py:91]: "Oleksii Kuchaiev  16 minutes ago: 
let’s merge - we have 2 approvals now. I’ll have a look at WER test and try to debug it"

File: nemo/core/classes/modelPT.py (hunk 13 above)
Comment (tkornuta-nvidia) [reply nemo/core/classes/modelPT.py:137]: Anyway, @okuchaiev  please, go ahead. 

Just leave the serialization to me, I really need to change several things down there to make it work for graphs.
